<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>HPCBio Single Cell Workshop • BuildACarpentriesWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="HPCBio Single Cell Workshop">
<meta property="og:description" content="BuildACarpentriesWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BuildACarpentriesWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CarpentriesWorkshopSyllabus.html">Building Carpentries-based Bioconductor Lessons</a>
    </li>
    <li>
      <a href="../articles/CellRanger_2020-10-30.html">HPCBio Single Cell Workshop</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://jdrnevich.github.io/BuildACarpentriesWorkshop/">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="CellRanger_2020-10-30_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>HPCBio Single Cell Workshop</h1>
                        <h4 class="author">Jenny Drnevich, Jessica Holmes and Lindsay Clark</h4>
            
            <h4 class="date">10/30/2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jdrnevich/BuildACarpentriesWorkshop/blob/master/vignettes/CellRanger_2020-10-30.Rmd"><code>vignettes/CellRanger_2020-10-30.Rmd</code></a></small>
      <div class="hidden name"><code>CellRanger_2020-10-30.Rmd</code></div>

    </div>

    
    
<div id="welcome---10x-single-cell-workshop" class="section level2">
<h2 class="hasAnchor">
<a href="#welcome---10x-single-cell-workshop" class="anchor"></a>Welcome - 10X Single Cell Workshop</h2>
<p>Jenny Drnevich, Instructor</p>
<p>Helpers: Jessica Holmes and Lindsay Clark</p>
<p>Workshop wiki: <a href="https://go.illinois.edu/10X-rnaseq" class="uri">https://go.illinois.edu/10X-rnaseq</a></p>
<p>Please download:</p>
<ul>
<li>This slide deck - 2020-10-30</li>
<li>Directory structure</li>
</ul>
<p>Make sure you have your hpcbioXX username and password from email</p>
</div>
<div id="introductions-and-poll" class="section level2">
<h2 class="hasAnchor">
<a href="#introductions-and-poll" class="anchor"></a>Introductions and Poll</h2>
<p>Instructor &amp; helpers</p>
<p>Participants</p>
<p>Poll 1: What computer OS are you using?</p>
<ol style="list-style-type: decimal">
<li>Linux</li>
<li>Mac</li>
<li>Windows</li>
</ol>
</div>
<div id="overview---part-1-on-biocluster" class="section level2">
<h2 class="hasAnchor">
<a href="#overview---part-1-on-biocluster" class="anchor"></a>Overview - Part 1 on Biocluster</h2>
<ol style="list-style-type: decimal">
<li><p>Brief review of bash, biocluster, reproducible research and file formats</p></li>
<li><p>Examine sequencing report and download data to biocluster</p></li>
<li>
<p>Go over Cell Ranger pipelines needed:</p>
<ul>
<li>
<code>cellranger mkgtf</code> and <code>cellranger mkref</code>
</li>
<li><code>cellranger count</code></li>
<li><code>cellranger aggr</code></li>
</ul>
</li>
<li><p>Examine Cell Ranger outputs</p></li>
<li><p>Explore using Loupe Browser</p></li>
</ol>
</div>
<div id="what-this-workshop-will-and-will-not-cover" class="section level2">
<h2 class="hasAnchor">
<a href="#what-this-workshop-will-and-will-not-cover" class="anchor"></a>What this workshop will and will not cover</h2>
<ol style="list-style-type: decimal">
<li><p>Will give hands-on demo of single cell gene expression libraries using <code>cellranger</code> and Seurat.</p></li>
<li><p>Will NOT do demo of ATAC or Visium libraries.</p></li>
<li><p>Will point out similarities of inputs to <code>cellranger-atac</code> and <code>spaceranger</code>, and learned pitfalls of <code>spaceranger</code>.</p></li>
</ol>
</div>
<div id="too-fast-too-slow" class="section level2">
<h2 class="hasAnchor">
<a href="#too-fast-too-slow" class="anchor"></a>Too Fast! Too Slow!</h2>
<p>It is impossible to have one best teaching speed for a mixed group of learners.</p>
<p>I will slow down my speaking speed (please remind me!)</p>
<p>To better serve a mixed group, for each task I have tried to add:</p>
<ul>
<li>Extra material for those done early</li>
<li>Expected material for the bulk of the class</li>
<li>Work-arounds for those who would need more time to finish</li>
</ul>
</div>
<div id="project-setup" class="section level2">
<h2 class="hasAnchor">
<a href="#project-setup" class="anchor"></a>Project setup</h2>
<ol style="list-style-type: decimal">
<li>Log in to the Biocluster via MobaXterm or terminal (Mac)</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> ssh hpcbioXX<span class="sc">@</span>biologin.igb.illinois.edu</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co">#enter password for your XX number - the cursor will NOT move!</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Use a worker node by doing an interactive srun for the classroom queue</li>
</ol>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> srun <span class="sc">-</span>p classroom <span class="sc">--</span>pty bash  </span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>Copy over a directory structure:</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cp <span class="sc">-</span>r <span class="sc">/</span>home<span class="sc">/</span>classroom<span class="sc">/</span>hpcbio<span class="sc">/</span>fa20<span class="sc">-</span>scRNA<span class="sc">/</span>scRNA<span class="dv">-2020</span><span class="sc">/</span> <span class="er">~/</span>  </span></code></pre></div>
<p>Remember to use tab-complete to get to the right place!</p>
</div>
<div id="where-are-you-are-on-biocluster" class="section level2">
<h2 class="hasAnchor">
<a href="#where-are-you-are-on-biocluster" class="anchor"></a>Where are you are on Biocluster?</h2>
<div class="red">
Login node vs. worker node?
</div>
<ol style="list-style-type: decimal">
<li>Look at beginning of command prompt to figure out:
<ol style="list-style-type: lower-alpha">
<li>
<code>[username@biologin</code> = login node</li>
<li>
<code>[username@compute-X-X</code> = worker node</li>
</ol>
</li>
<li>Login: scheduler commands, submit scripts, <strong>INTERNET</strong>
</li>
<li>Worker: interactive working, run small jobs</li>
</ol>
<div class="red">
Place in directory structure?
</div>
<ol style="list-style-type: decimal">
<li>Look at end of command prompt to figure out:
<ol style="list-style-type: lower-alpha">
<li><code>currentDirName]$</code></li>
<li>
<code>~]$</code> means your home directory</li>
</ol>
</li>
<li>
<code>pwd</code> will give full path to your current directory</li>
<li>Keep in mind relative vs. full path when writing scripts!</li>
</ol>
</div>
<div id="directory-structure" class="section level2">
<h2 class="hasAnchor">
<a href="#directory-structure" class="anchor"></a>Directory structure</h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cd <span class="sc">~</span><span class="er">/</span>scRNA<span class="dv">-2020</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> pwd</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="sc">/</span>home<span class="sc">/</span>a<span class="sc">-</span>m<span class="sc">/</span>hpcinstru05<span class="sc">/</span>scRNA<span class="dv">-2020</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> tree</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>.</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>├── data</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>│   ├── genome</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>│   └── raw<span class="sc">-</span>seq</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>├── results</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>│   └── cellranger_1M</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>└── src</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>    ├── JobScriptStart.sh</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>    ├── pack_CR_results.sh</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>    ├── slurm<span class="sc">-</span>out</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>    └── uncompress<span class="sc">-</span>tbz.sh</span></code></pre></div>
<p>NOTE: <code>tree</code> does not work on login node</p>
</div>
<div id="bash-command-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#bash-command-usage" class="anchor"></a>Bash command usage</h2>
<p><code>commandName [options] [argument1] [argument2]...</code></p>
<ul>
<li>First word of new line interpreted as a command/program to run; everything is separated by spaces.</li>
<li>Options start with <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code> or <code>--</code> and can be in any order. To set a value for an option, do it one space after: <code>-p 4</code>
</li>
<li>Arguments often are input files to command. They must in in order and after all options</li>
<li>The entire command-options-arguments must be on the same line; to split to multiply lines for clarity, end each line with <code>\</code>.</li>
</ul>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> srun <span class="sc">-</span>p classroom <span class="sc">--</span>pty bash  </span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cp <span class="sc">-</span>r <span class="sc">/</span>home<span class="sc">/</span>classroom<span class="sc">/</span>hpcbio<span class="sc">/</span>fa20<span class="sc">-</span>scRNA<span class="sc">/</span>scRNA<span class="dv">-2020</span><span class="sc">/</span> <span class="er">~/</span>  </span></code></pre></div>
<p>Poll Question 2</p>
</div>
<div id="reproducible-vs--replicable" class="section level2">
<h2 class="hasAnchor">
<a href="#reproducible-vs--replicable" class="anchor"></a>Reproducible vs. Replicable</h2>
<div class="red">
<strong>Reproducibility</strong>
</div>
<p>is the ability to re-compute the analytic results of a study given the original data set and a description of the steps taken in the original analysis. <a href="http://www.sciencemag.org/content/334/6060/1226.abstract">Peng 2011, Science</a></p>
<div class="red">
<strong>Replicability</strong>
</div>
<p>is the likelihood that a new study on the same scientific question will produce a result consistent with the original study. <a href="http://onlinelibrary.wiley.com/doi/10.1002/per.1919/abstract">Asendorph 2013 European J. of Personality</a></p>
</div>
<div id="reproducible-data-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#reproducible-data-analysis" class="anchor"></a>Reproducible Data Analysis</h2>
<ul>
<li>You should have a (virtual) computational notebook like you have a lab notebook</li>
<li>Every detail of the data analysis needs to be recorded so that you or anyone else could reproduce the end results
<ul>
<li>Software, versions, options specified</li>
<li>All data manipulations/normalizations/transformations</li>
<li>Exact statistical methods used</li>
<li>How each figure/table was calculated and made</li>
</ul>
</li>
</ul>
</div>
<div id="tips-for-reproducibility" class="section level2">
<h2 class="hasAnchor">
<a href="#tips-for-reproducibility" class="anchor"></a>Tips for reproducibility</h2>
<ul>
<li>Avoid using programs that do not record all data manipulations that were done (e.g., Excel, point-and-click pop-up boxes; Loupe Browser?!)</li>
<li>Use written scripts instead of typing in commands directly at the prompt (Linux, R, Python, etc.)</li>
<li>Use helper programs that allow you to integrate verbal explanations with actual codes and computed output (Rmarkdown, iPython notebook, etc.)</li>
<li>Collate bare-minimum explanations, scripts and codes for each publication for inclusion in supplementary materials or at least “available upon request”</li>
</ul>
</div>
<div id="reproducibility-for-biocluster" class="section level2">
<h2 class="hasAnchor">
<a href="#reproducibility-for-biocluster" class="anchor"></a>Reproducibility for Biocluster</h2>
<ul>
<li>Log in to biocluster a second (or third) time</li>
<li>Change to <code>src</code> directory and create a file to take notes:</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> srun <span class="sc">-</span>p classroom <span class="sc">--</span>pty bash</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cd <span class="sc">~</span><span class="er">/</span>scRNA<span class="dv">-2020</span><span class="sc">/</span>src  </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> nano Notes<span class="sc">-</span>scRNA_biocluster.txt</span></code></pre></div>
<p>Don’t forget to save often with <code>ctrl-O</code></p>
</div>
<div id="file-formats-review" class="section level2 smaller">
<h2 class="hasAnchor">
<a href="#file-formats-review" class="anchor"></a>File formats review</h2>
<ol style="list-style-type: decimal">
<li>Fasta (.fasta or .fna or .fa)
<ol style="list-style-type: lower-alpha">
<li>Contains sequences</li>
<li>Usually genome or transcriptome</li>
</ol>
</li>
<li>Fastq (.fastq)
<ol style="list-style-type: lower-alpha">
<li>Contains sequences + quality scores</li>
<li>Usually output from sequencer</li>
</ol>
</li>
<li>Gene models (.gff or .gtf)
<ol style="list-style-type: lower-alpha">
<li>Contains locations of features on genome</li>
<li>Tab-delimited text file with 9 columns
<ol style="list-style-type: decimal">
<li>3rd column is feature (region, gene, mRNA, exon, etc.)</li>
<li>9th column has attributes; important ones:
<ol style="list-style-type: lower-alpha">
<li>gene_id</li>
<li>transcript_id</li>
<li>gene_biotype or gbkey</li>
</ol>
</li>
</ol>
</li>
</ol>
</li>
</ol>
</div>
<div id="practice-data-set" class="section level2">
<h2 class="hasAnchor">
<a href="#practice-data-set" class="anchor"></a>Practice Data Set</h2>
<ul>
<li>Generated Aug 2019, v3 chemistry, <code>cellranger mkfastq</code> 3.1</li>
<li>2 mouse brain samples: Control and Treated (fake!)</li>
<li>Full samples have ~750M reads each</li>
<li>Practice sets of 1M reads each</li>
<li>Download sequencing report from <a href="https://go.illinois.edu/10X-rnaseq">box</a> Week 1 Materials:
<ul>
<li>Note names of tar file/s</li>
<li>Info on versions for Chromium kit, sequencer and Cell Ranger</li>
<li>username and password for download (if ftpserver)</li>
</ul>
</li>
</ul>
</div>
<div id="sample-dual-vs--single-indexing" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-dual-vs--single-indexing" class="anchor"></a>Sample dual vs. single indexing</h2>
<p><a href="https://assets.ctfassets.net/an68im79xiti/Licpd2PiHP4hrHKDpjO89/2779c006e6317ed9ca724635b32e14e9/CG000325_TechNote_ChromiumNextGEMSingle_Cell_3___v3.1_Dual_Index_Rev_A.pdf">10X Technical Note</a></p>
<ul>
<li>Older 10X libraries used only a single sample index (i5), used to demultiplex reads to each sample.</li>
<li>“Index hopping” can occur in 0.1-2% of reads, resulting in assignment to the wrong sample.</li>
<li>New dual indexing uses unique combinations of i5 and i7; reads with unexpected i5-i7 combinations are thrown out.</li>
</ul>
</div>
<div id="sets-of--fastq-files-per-sample" class="section level2">
<h2 class="hasAnchor">
<a href="#sets-of--fastq-files-per-sample" class="anchor"></a>Sets of .fastq files per sample:</h2>
<p>.fastq.gz will come in sets of 3 or sets of 4 per sample:</p>
<p>Set of 3:</p>
<ol style="list-style-type: decimal">
<li>R1 - 10X bead barcode + UMI</li>
<li>R2 - actual read sequence</li>
<li>I1 - sample i5 index</li>
</ol>
<p>Set of 4: same as above + I2 - sample i7 index</p>
<p>After sample demultiplexing, the I1/I2 fastqs are not needed by the user!</p>
<p>More than one set per sample: aliquots were run in different sequencing lanes (L001, L002, etc.).</p>
</div>
<div id="poll-3--how-do-you-usually-download-files" class="section level2">
<h2 class="hasAnchor">
<a href="#poll-3--how-do-you-usually-download-files" class="anchor"></a>Poll 3. How do you usually download files?</h2>
<p>Please mark all the ways you have used to download files to your computer:</p>
<ol style="list-style-type: decimal">
<li>Through a browser by clicking on a link</li>
<li>Through software like CyberDuck, WinSCP, Filezilla, etc.</li>
<li>Via the command line using scp, curl or wget</li>
<li>Other method - put in chat</li>
</ol>
</div>
<div id="switch-from-ftpserver-to-posting-server" class="section level2">
<h2 class="hasAnchor">
<a href="#switch-from-ftpserver-to-posting-server" class="anchor"></a>Switch from ftpserver to posting server</h2>
<ul>
<li>DNA Services is switching from a ftpserver to a new Amazon S3 posting server</li>
<li>The ftpserver has a <a href="https://help.igb.illinois.edu/Globus">Globus endpoint</a> you can use to download to your own computer or directly to Biocluster.</li>
<li>The new posting server will not have a Globus endpoint.</li>
<li>We are working with DNA Services to improve the posting server download instructions; expect complete switch by January 2021.</li>
</ul>
</div>
<div id="if-your-data-is-on-the-ftpserver" class="section level2">
<h2 class="hasAnchor">
<a href="#if-your-data-is-on-the-ftpserver" class="anchor"></a>If your data is on the ftpserver:</h2>
<ol style="list-style-type: decimal">
<li>Your sequencing report will have a username &amp; password and download instructions.</li>
<li>Any of the download methods will work to get a copy on your personal/lab computer. <em>NOTE: this make take hours!</em>
</li>
<li>The Globus method is recommended to transfer a copy directly from the ftpserver to biocluster. See CNRG’s <a href="https://help.igb.illinois.edu/Globus">Globus guide</a>.</li>
<li>Data on the ftpserver will be available for (at least) 30 days.</li>
</ol>
<div class="red">
<p>NOTE: You need to make two backup copies of your sequencing data. A copy on Biocluster does <em>NOT</em> count. Use CNRG’s <a href="https://help.igb.illinois.edu/Active_Archive">Archiving service</a> instead.</p>
</div>
<div id="if-your-data-is-on-the-posting-server" class="section level2">
<h2 class="hasAnchor">
<a href="#if-your-data-is-on-the-posting-server" class="anchor"></a>If your data is on the posting server:</h2>
<ol style="list-style-type: decimal">
<li>You will receive download instructions in a separate email from the sequencing report.</li>
<li>The clickable link will start a browser download to your computer. <em>NOTE: this make take hours!</em>
</li>
<li>You can also use curl or wget on your own machine (<em>video instructions will be made in the next couple of months</em>) or on <strong>biocluster</strong> to download a copy.</li>
<li>The links for the posting server expire after <strong>7 days</strong> but can be re-issued upon request.</li>
</ol>
<div class="red">
<p>NOTE: You need to make two backup copies of your sequencing data. A copy on Biocluster does <em>NOT</em> count. Use CNRG’s <a href="https://help.igb.illinois.edu/Active_Archive">Archiving service</a> instead.</p>
</div>
<div id="downloading-data-from-posting-server" class="section level2">
<h2 class="hasAnchor">
<a href="#downloading-data-from-posting-server" class="anchor"></a>Downloading data from posting server</h2>
<p>I sent you an example email:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>The <span class="fu">links</span> (URLs) will expire on <span class="dv">2020-11-05</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>File<span class="sc">:</span> Project_PIname_10X.tar.bz2 </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>URL<span class="sc">:</span> https<span class="sc">:</span><span class="er">//</span>posting.biotech.illinois.edu<span class="sc">/</span>posting<span class="sc">/</span>hpcbio<span class="sc">/</span>Project_PIname_10X.tar.bz2?x<span class="sc">-</span>email<span class="ot">=</span>cjfields<span class="sc">%40illinois.edu&amp;X-Amz-Algorithm=AWS4-HMAC-SHA256&amp;X-Amz-Credential=posting%</span>2F20201029<span class="sc">%2Fus-east-1%</span>2Fs3%2Faws4_request<span class="sc">&amp;</span>X<span class="sc">-</span>Amz<span class="sc">-</span>Date<span class="ot">=</span>20201029T131333Z<span class="sc">&amp;</span>X<span class="sc">-</span>Amz<span class="sc">-</span>Expires<span class="ot">=</span><span class="dv">604800</span><span class="sc">&amp;</span>X<span class="sc">-</span>Amz<span class="sc">-</span>SignedHeaders<span class="ot">=</span>host<span class="sc">&amp;</span>X<span class="sc">-</span>Amz<span class="sc">-</span>Signature<span class="ot">=</span>6794aa1cc12134744b4ca06e54c1b1e0c590800ccab59bbdaebb1a95d19cb39a </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>md5sum<span class="sc">:</span> 310c8cc625f2bf14484277cec499c0be </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>________________________________________</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>To download the file via the command line</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>curl <span class="sc">-</span>L <span class="sc">-</span>o FILENAME <span class="st">"https://....."</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>or</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>wget <span class="sc">-</span>O FILENAME <span class="st">"https://....."</span></span></code></pre></div>
<p>If you click on the link in the email, it will start download to <em>your computer</em> using your default browser. This may take <strong>hours</strong> and your browser needs to stay open and your computer connected to the internet.</p>
</div>
<div id="download-practice-data-using-curl" class="section level2">
<h2 class="hasAnchor">
<a href="#download-practice-data-using-curl" class="anchor"></a>Download practice data using curl</h2>
<p>Exit the worker node to get back to the login node, and make sure you are in the directory you want:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> exit  </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co">#check prompt to make sure now on login node</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cd <span class="sc">~</span><span class="er">/</span>scRNA<span class="dv">-2020</span><span class="sc">/</span>data<span class="sc">/</span>raw<span class="sc">-</span>seq</span></code></pre></div>
<p>Start typing the curl command, putting in the correct FILENAME:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> curl <span class="sc">-</span>L <span class="sc">-</span>o Project_PIname_10X.tar.bz2 <span class="co">#but don't hit enter yet...</span></span></code></pre></div>
<p>In the email, right-click on the clickable URL and say “Copy hyperlink”. Then finish the curl command by typing in <code>"</code>, pasting the link, then finishing with <code>"</code>, then hit enter.</p>
<p>The link must be within quotes!</p>
</div>
<div id="backup-for-workshop" class="section level2">
<h2 class="hasAnchor">
<a href="#backup-for-workshop" class="anchor"></a>Backup for workshop</h2>
<p>If you are having trouble downloading from the posting server, copy from:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cd <span class="sc">~</span><span class="er">/</span>scRNA<span class="dv">-2020</span><span class="sc">/</span>data<span class="sc">/</span>raw<span class="sc">-</span>seq</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> ls <span class="sc">-</span>lh  <span class="co">#if do not see .bz2 file, do:</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cp <span class="sc">/</span>home<span class="sc">/</span>classroom<span class="sc">/</span>hpcbio<span class="sc">/</span>fa20<span class="sc">-</span>scRNA<span class="sc">/</span>1M_example<span class="sc">/</span>data<span class="sc">/</span>raw<span class="sc">-</span>seq<span class="sc">/</span>Project_PIname_10X.tar.bz2 .<span class="sc">/</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="er">$</span> ls <span class="sc">-</span>lh</span></code></pre></div>
</div>
<div id="check-md5sum" class="section level2">
<h2 class="hasAnchor">
<a href="#check-md5sum" class="anchor"></a>Check md5sum</h2>
<p>There is potential for errors or corruption when downloading/transferring large files. MD5 hashing gives a unique digital fingerprint of a file and any difference in the file will change the hash code. The posting server e-mail gives the md5sum for the file, so we can check to make sure it did not get corrupted.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Get back on worker node</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> srun <span class="sc">--</span>pty <span class="sc">-</span>p classroom bash</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> md5sum Project_PIname_10X.tar.bz2</span></code></pre></div>
<p>Compare the md5sum generated with the one in the email.</p>
</div>
<div id="extract-data-from--tar-bz2" class="section level2">
<h2 class="hasAnchor">
<a href="#extract-data-from--tar-bz2" class="anchor"></a>Extract data from .tar.bz2</h2>
<p>The individual .fastq.gz files have been put in a tar archive and the compressed using bzip2. To extract them, first use the method given in the sequencing report:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> tar <span class="sc">-</span>xvf filename.bz2</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> ls <span class="sc">-</span>lh</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> tree</span></code></pre></div>
<div class="red">
Note: full size files may take hours to extract with this method. You <em>must</em> remain logged in if doing interactively
</div>
<p><br></p>
<p>Alternative 1: use <a href="https://linuxize.com/post/how-to-use-linux-screen/">screen</a> functionality to keep your session going even if you log out.</p>
<p>Alternative 2: Submit as job script, either with <code>tar -xvf</code> code or by…</p>
</div>
<div id="parallel-extraction-using-script" class="section level2">
<h2 class="hasAnchor">
<a href="#parallel-extraction-using-script" class="anchor"></a>Parallel extraction using script</h2>
<p>See example script:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cat ..<span class="sc">/</span>..<span class="sc">/</span>src<span class="sc">/</span>uncompress<span class="sc">-</span>tbz.sh</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co">#If you want to try to run this, rename the other extraction:</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> mv Project_PIname_10X Project_PIname_10X_old</span></code></pre></div>
<p>This script is not meant to be changed internally but have the .tar.bz passed as a varible in the sbatch call. The script has a comment showing how to run it:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> sbatch <span class="sc">--</span>export<span class="ot">=</span>TBZ<span class="ot">=</span>Martinis.<span class="fl">201875.</span>bz2 uncompress<span class="sc">-</span>tbz.sh</span></code></pre></div>
<p>Breakout rooms: figure out how to run this batch script on Project_PIname_10X.tar.bz. Warning: must point to .bz2 file and .sh script correctly!</p>
<p><strong>Advanced</strong>: Can you figure out how to use screen to open a screen session, detach it, then re-attach it? See basic steps at end of webpage.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> ls <span class="sc">-</span>lh</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cat uncompress.out</span></code></pre></div>
</div>
<div id="remove--tar-bz-and-_old-directory-if-made" class="section level2">
<h2 class="hasAnchor">
<a href="#remove--tar-bz-and-_old-directory-if-made" class="anchor"></a>Remove .tar.bz (and _old directory if made)</h2>
<p>You pay for storage on biocluster, so you do not need the .tar.bz after extracting files. Remove it:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> rm Project_PIname_10X.tar.bz2 <span class="co"># rm -rf Project_PIname_10X_old/</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Final directory structure:</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> tree</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>.</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>├── Project_PIname_10X</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>│   ├── Control</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>│   │   ├── Control_1M_L001_I1_001.fastq.gz</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>│   │   ├── Control_1M_L001_R1_001.fastq.gz</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>│   │   └── Control_1M_L001_R2_001.fastq.gz</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>│   └── Treated</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>│       ├── Treated_1M_L001_I1_001.fastq.gz</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>│       ├── Treated_1M_L001_R1_001.fastq.gz</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>│       └── Treated_1M_L001_R2_001.fastq.gz</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>├── uncompress<span class="fl">-6259647.</span>out</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span> directories, <span class="dv">7</span> files</span></code></pre></div>
</div>
<div id="cell-ranger-pipelines-for-sc-gene-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-ranger-pipelines-for-sc-gene-expression" class="anchor"></a><a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> pipelines for sc gene expression</h2>
<ul>
<li>
<a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/mkfastq"><code>cellranger mkfastq</code></a>
<ul>
<li>Demultiplexes samples</li>
<li>Run by our Sequencing Center, not user</li>
</ul>
</li>
<li>
<a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/count"><code>cellranger count</code></a>
<ul>
<li>Run by user</li>
<li>Separately for each sample</li>
</ul>
</li>
<li>
<a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/aggregate"><code>cellranger aggr</code></a>
<ul>
<li>Puts &gt; 1 sample in same .cloupe file</li>
<li>Only normalization is sub-sampling</li>
</ul>
</li>
<li>
<a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/reanalyze"><code>cellranger reanalyze</code></a>
<ul>
<li>Used to modify parameters of previous run</li>
</ul>
</li>
</ul>
</div>
<div id="custom-reference" class="section level2">
<h2 class="hasAnchor">
<a href="#custom-reference" class="anchor"></a>Custom reference?</h2>
<p><a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_mr" class="uri">https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_mr</a></p>
<ul>
<li>Pre-built 10X not updated frequently (last done July 2020, Ensembl release 98; current release is 101)</li>
<li>Only available for human (GRCh38 &amp; hg19) and/or mouse (GRCm38)</li>
<li>If you want to use a different gene model set, add genes/species, or non-human/mouse organism, you will have to make a custom reference.</li>
</ul>
</div>
<div id="gtf-must-be-compatible-with-star" class="section level2">
<h2 class="hasAnchor">
<a href="#gtf-must-be-compatible-with-star" class="anchor"></a>GTF must be compatible with STAR</h2>
<ul>
<li>Must be .gtf, not .gff</li>
<li>Feature type (column 3) of “exon”</li>
<li>Attributes (column 9) of “gene_id” and “transcript_id”</li>
<li>Ensembl/Gencode gtf files generally compatible.</li>
<li>NCBI gtf files should be compatible, but gff files are NOT without additional manipulation.</li>
<li>See <a href="https://go.illinois.edu/introR">Intro to Bioconductor workshop</a>, Session 2 for R codes to transform .gff file to .gtf with correct attributes.</li>
</ul>
</div>
<div id="download-newest-ensembl-gtf-only" class="section level2 smaller">
<h2 class="hasAnchor">
<a href="#download-newest-ensembl-gtf-only" class="anchor"></a>Download newest Ensembl gtf only</h2>
<p><a href="https://uswest.ensembl.org/Mus_musculus/Info/Index">Ensembl’s mouse genome page</a>. Can you follow 10X’s instructions?</p>
<p>“Navigate to the Gene annotation section of the Ensembl website and click on the Download GTF link. This takes you to an ftp site with a list of GTF files available. Select the file called Danio_rerio.GRCz11.99.chr.gtf.gz. This is the GTF annotation file for this species. <strong>All species in Ensembl have similar files available to download.</strong> For more information on the GTF files in Ensembl, click on the README file.”</p>
<p>Breakout rooms:</p>
<ol style="list-style-type: decimal">
<li>Download the mouse .gtf file to <code>~/scRNA-2020/data/genome</code> using <code>wget</code>. Hint: where on biocluster do you <em>have</em> to be to run <code>wget</code> (internet access)?</li>
<li>Extract the .gtf file. Hint: where on biocluster <em>should</em> you be to run <code>gunzip</code>?</li>
<li>
<strong>Need help?</strong> See codes next slide</li>
<li>
<strong>Advanced?</strong> Download the gtf.gz file again. Ensembl’s ftp site also has CHECKSUMS. Make sure your file downloaded fine, but use <code>sum</code> instead of <code>md5sum</code>.</li>
</ol>
</div>
<div id="answer-codes" class="section level2 smaller">
<h2 class="hasAnchor">
<a href="#answer-codes" class="anchor"></a>Answer codes</h2>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># exit to login node</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> exit</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Navigate to directory </span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cd <span class="sc">~</span><span class="er">/</span>scRNA<span class="dv">-2020</span><span class="sc">/</span>data<span class="sc">/</span>genome</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Download using wget</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> wget ftp<span class="sc">:</span><span class="er">//</span>ftp.ensembl.org<span class="sc">/</span>pub<span class="sc">/</span>release<span class="dv">-101</span><span class="sc">/</span>gtf<span class="sc">/</span>mus_musculus<span class="sc">/</span>Mus_musculus.GRCm38.<span class="fl">101.</span>chr.gtf.gz</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Get back on worker node</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> srun <span class="sc">-</span>p classroom <span class="sc">--</span>pty bash</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Bonus: checksum should be 11872 32653</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> sum Mus_musculus.GRCm38.<span class="fl">101.</span>chr.gtf.gz</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the file</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> gunzip Mus_musculus.GRCm38.<span class="fl">101.</span>chr.gtf.gz</span></code></pre></div>
</div>
<div id="filter-out-transcriptsgenes" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-out-transcriptsgenes" class="anchor"></a>Filter out transcripts/genes?</h2>
<ul>
<li>10X recommends filtering gtf to keep only certain kinds of genes</li>
<li>Explanation as to why somewhat lacking…</li>
<li>But likely does not make that much difference</li>
<li>10X’s <code>cellranger mkgtf</code> <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_mr#filter">documentation</a>
</li>
<li>10X’s current <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/release-notes/build#mm10_2020A">build steps</a> for mouse reference uses some very sophisticated bash scripting!</li>
</ul>
</div>
<div id="run-cellranger-mkgtf" class="section level2">
<h2 class="hasAnchor">
<a href="#run-cellranger-mkgtf" class="anchor"></a>Run <code>cellranger mkgtf</code>
</h2>
<p>The <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_mr#filter">tutorial</a> shows us an example of code to run, but not how to successfully run it on Biocluster. Try copying, modifying and pasting the <code>cellranger mkgtf</code> code to run it and see what happens…</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>bash<span class="sc">:</span> cellranger<span class="sc">:</span> command not found</span></code></pre></div>
<p>On Biocluster, additional software needs to have its module loded first:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> module avail cellranger</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> module load cellranger<span class="sc">/</span><span class="dv">4</span>.<span class="fl">0.0</span></span></code></pre></div>
<p>Then could run code interactively, which could take ~10 min, or could run in job script…</p>
</div>
<div id="job-script-review" class="section level2 smaller">
<h2 class="hasAnchor">
<a href="#job-script-review" class="anchor"></a>Job Script review</h2>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cd <span class="sc">~</span><span class="er">/</span>scRNA<span class="dv">-2020</span><span class="sc">/</span>src</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cat JobScriptStart.sh</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash                               #shebang line; use bash interpreter</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -N 1                              #number of computer nodes to reserve</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -n 8                             #number of processors/threads/cpus to reserve</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mem=50G                         #how much memory to reserve</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -p classroom                      #which partition to run on</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mail-user=MyEmail@illinois.edu  #where to send emails</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mail-type=</span><span class="re">END</span><span class="co">,FAIL              #when to send emails</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -J MyJobName                      #make a descriptive job name</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --output=OutFileName-%A.out       #put any stout in this file name; %A adds on job id</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH -D /home/a-m/hpcbioXX/scRNA-2020/src/slurm-out #see description in .sh</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a><span class="do">### Load Modules</span></span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>module load cellranger<span class="sc">/</span><span class="dv">4</span>.<span class="fl">0.0</span></span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a><span class="do">### Take input memory and put 90% in new variable that can be used in</span></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a><span class="do">### cellranger calls as --localmem=$memG90</span></span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>memG90<span class="ot">=</span><span class="st">`</span><span class="at">awk "BEGIN { printf </span><span class="sc">\"</span><span class="at">%.0f</span><span class="sc">\n\"</span><span class="at">, $SLURM_MEM_PER_NODE /1000*0.9}"</span><span class="st">`</span></span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a><span class="do">### Change to desired directory</span></span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a>cd <span class="sc">~</span><span class="er">/</span>scRNA<span class="dv">-2020</span></span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a><span class="do">### Run commands</span></span></code></pre></div>
</div>
<div id="copy-and-modify-job-script-start" class="section level2 smaller">
<h2 class="hasAnchor">
<a href="#copy-and-modify-job-script-start" class="anchor"></a>Copy and modify job script start</h2>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cp JobScriptStart.sh CR_mkgtf.sh</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> nano CR_mkgtf.sh</span></code></pre></div>
<p>In #SBATCH header:</p>
<ol style="list-style-type: decimal">
<li>change <code>-n</code> to 1</li>
<li>Change email address</li>
<li>Change job name to <code>mkgtf</code>
</li>
<li>Change OutFileName to <code>mkgtf-%A.out</code>
</li>
<li>Change hpcbioXX in <code>-D</code> to your number; <strong>must use full path</strong> to your directory or else job disappear with no errors.</li>
</ol>
<p>In body of script:</p>
<ol start="5" style="list-style-type: decimal">
<li>Change desired directory to <code>cd ~/scRNA-2020/data/genome</code>
</li>
<li>Add in corrected <code>cellranger mkgtf</code> command</li>
<li>Save, exit then submit:</li>
</ol>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> sbatch CR_mkgtf.sh</span></code></pre></div>
</div>
<div id="did-my-job-run-ok" class="section level2">
<h2 class="hasAnchor">
<a href="#did-my-job-run-ok" class="anchor"></a>Did my job run ok?</h2>
<p>Once you submit your job, check on it by:</p>
<ol style="list-style-type: decimal">
<li><code>squeue -u hpcbioXX</code></li>
<li>Look at the .out file in <code>slurm-out/</code> directory</li>
<li>Check for an email from SLURM</li>
<li>See if the expected output files are present</li>
</ol>
</div>
<div id="do-we-have-time-to-run-cellranger-mkref" class="section level2">
<h2 class="hasAnchor">
<a href="#do-we-have-time-to-run-cellranger-mkref" class="anchor"></a>Do we have time to run <code>cellranger mkref</code>?</h2>
<ul>
<li>It takes over 30 min, so good if we can get it started by the end of today.</li>
<li>If not, it only needs to be done once per genome/gtf
<ul>
<li>You can choose to try it as homework - see next slides</li>
<li>You can also just use one I have already made - details next week</li>
</ul>
</li>
</ul>
<p>If ending now, make sure to save your notes and exit out of all connections!</p>
<p>If continuing, need to download genome (<em>“.dna.primary_assembly.fa.gz”</em>) from <a href="https://uswest.ensembl.org/Mus_musculus/Info/Index">Ensembl</a> to <code>data/genome</code> like we did for the .gtf. Can you follow <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_mr#getfiles">10X’s instructions</a>?</p>
</div>
<div id="warning-about-cores-and-memory" class="section level2">
<h2 class="hasAnchor">
<a href="#warning-about-cores-and-memory" class="anchor"></a>Warning about cores and memory</h2>
<p><code>cellranger</code> runs on Biocluster in <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/job-submission-mode">Job Submission Mode</a>. Per (old) 10X:</p>
<p>“By default, cellranger uses all available cores and 90% of detected memory. This behavior may be undesirable in a shared environment with multiple concurrent users and tasks. It is strongly recommended to run cellranger with <code>--localcores</code> and <code>--localmem</code> to specify resource usage upper bounds. In practice, there is negligible return in allocating more than 32 cores or 256G to the pipeline. Check the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/overview/system-requirements">system requirements</a> page for more information on how resource allocations affect pipeline runtime.”</p>
<p><em>Always</em> set <code>--localcores</code> and <code>--localmem</code> if using <code>cellranger count</code>, <code>cellranger aggr</code> or <code>cellranger reanalyze</code></p>
<p>(<code>cellranger mkref</code> uses <code>--nthreads</code> and <code>--memgb</code> instead)</p>
<p>Check Biocluster’s <a href="https://help.igb.illinois.edu/Biocluster2#Cluster_Specifications">queue specifications</a></p>
</div>
<div id="make-cellranger-mkref-script" class="section level2">
<h2 class="hasAnchor">
<a href="#make-cellranger-mkref-script" class="anchor"></a>Make <code>cellranger mkref</code> script</h2>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cd ~/scRNA-2020/src</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> cp CR_mkgtf.sh CR_mkref.sh</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> nano CR_mkref.sh</span></code></pre></div>
<ol style="list-style-type: decimal">
<li><p>Change <code>-n</code> to 8</p></li>
<li><p>Change job name to <code>mkref</code></p></li>
<li><p>Change OutFileName to mkref-%A.out</p></li>
<li>
<p>Remove <code>cellranger mkgtf</code> command and put in correct <code>cellranger mkref</code> command:</p>
<ol style="list-style-type: lower-alpha">
<li>Copy/paste/modify the <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/tutorial_mr#mkrefsetup">tutorial example</a>.</li>
<li>Name the genome <code>GRCm38_ensembl101</code>
</li>
<li>Add extra arguments: <code>--nthreads=$SLURM_NTASKS</code> and <code>--memgb=$memG90</code>
</li>
</ol>
</li>
</ol>
</div>
<div id="save-and-submit-the-script" class="section level2">
<h2 class="hasAnchor">
<a href="#save-and-submit-the-script" class="anchor"></a>Save and submit the script</h2>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span> sbatch CR_mkref.sh</span></code></pre></div>
<p>Once you submit your job, check on it by:</p>
<ol style="list-style-type: decimal">
<li><code>squeue -u hpcbioXX</code></li>
<li>Look at the .out file in <code>slurm-out/</code> directory</li>
<li>Check for an email from SLURM</li>
<li>See if the expected output files are present</li>
</ol>
</div>
<div id="similarites-to-spaceranger-and-cellranger-atac-custom-references" class="section level2">
<h2 class="hasAnchor">
<a href="#similarites-to-spaceranger-and-cellranger-atac-custom-references" class="anchor"></a>Similarites to <code>spaceranger</code> and <code>cellranger-atac</code> custom references</h2>
<p>Spaceranger’s <a href="https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/advanced/references"><code>mkgtf/mkref</code></a> pretty much identical to <code>cellranger mkgtf/mkref</code>. (In fact, 10X told me a few months ago that a reference made with cellranger could be used for spaceranger).</p>
<p><code>cellranger-atac</code> does not say to filter with <code>mkgtf</code>. There is a <a href="https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/advanced/references">`mkref’</a> but they <strong>STRONGLY</strong> suggest using their pre-built references as they have info on regulatory regions.</p>
<p><code>cellranger-atac mkref</code> is run in a different format, with a <a href="https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/advanced/references#config">configuration file</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jenny Drnevich, Charlotte Soneson, Toby Hodges, Laurent Gatto.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </div>
</div>
</body>
</html>
